plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/HudiReadOptimizedFileLister.java

```diff
@@ -0,0 +1,220 @@
+        implements HudiFileLister
+{
+    private static final Logger log = Logger.get(HudiReadOptimizedFileLister.class);
+
+    final HoodieMetadataConfig metadataConfig;
+    final HoodieEngineContext engineContext;
+    final HoodieTableMetaClient metaClient;
+    final HudiTableHandle tableHandle;
+    final HiveMetastore hiveMetastore;
+    final Table hiveTable;
+    final SchemaTableName tableName;
+    final List<HiveColumnHandle> partitionColumnHandles;
+    final boolean useMetastoreForPartitions;
+    final HoodieTableFileSystemView fileSystemView;
+    final TupleDomain<String> partitionKeysFilter;
+    final List<Column> partitionColumns;
+
+    private List<String> hivePartitionNames;
+
+    public HudiReadOptimizedFileLister(
+            HoodieMetadataConfig metadataConfig,
+            HoodieEngineContext engineContext,
+            HudiTableHandle tableHandle,
+            HoodieTableMetaClient metaClient,
+            HiveMetastore hiveMetastore,
+            Table hiveTable,
+            List<HiveColumnHandle> partitionColumnHandles,
+            boolean useMetastoreForPartitions)
+    {
+        this.metadataConfig = metadataConfig;
+        this.engineContext = engineContext;
+        this.tableHandle = tableHandle;
+        this.tableName = tableHandle.getSchemaTableName();
+        this.metaClient = metaClient;
+        this.hiveMetastore = hiveMetastore;
+        this.hiveTable = hiveTable;
+        this.partitionColumnHandles = partitionColumnHandles;
+        this.useMetastoreForPartitions = useMetastoreForPartitions;
+        this.fileSystemView = FileSystemViewManager.createInMemoryFileSystemView(engineContext, metaClient, metadataConfig);
+        this.partitionKeysFilter = MetastoreUtil.computePartitionKeyFilter(partitionColumnHandles, tableHandle.getPartitionPredicates());
+        this.partitionColumns = hiveTable.getPartitionColumns();
+    }
+
+    @Override
+    public List<HudiPartitionInfo> getPartitionsToScan()
+    {
+        HoodieTimer timer = new HoodieTimer().startTimer();
+
+        List<HudiPartitionInfo> allPartitionInfoList = null;
+
+        if (!useMetastoreForPartitions) {
+            try {
+                // Use relative partition path and other context to construct
+                // InternalHudiPartitionInfo instances
+                PartitionValueExtractor partitionValueExtractor = partitionColumns.isEmpty()
+                        ? new NonPartitionedExtractor()
+                        : inferPartitionValueExtractorWithHiveMetastore();
+                List<String> relativePartitionPathList = partitionColumns.isEmpty()
+                        ? Collections.singletonList("")
+                        : TimelineUtils.getPartitionsWritten(metaClient.getActiveTimeline());
+                allPartitionInfoList = relativePartitionPathList.stream()
+                        .map(relativePartitionPath ->
+                                buildHudiPartitionInfo(
```

Comment:
It would be better to inline this method, since the `useMetastoreForPartitions` field is a constant here.